pkg/util/stop/stop_test: TestStopperRunAsyncTaskTracing failed

pkg/util/stop/stop_test.TestStopperRunAsyncTaskTracing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5663519?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5663519?buildTab=artifacts#/) on master @ [d239a2fbf9cc3c01d2708811824c54a8091b29ee](https://github.com/cockroachdb/cockroach/commits/d239a2fbf9cc3c01d2708811824c54a8091b29ee):


```
=== RUN   TestStopperRunAsyncTaskTracing
    stopper_test.go:747: 
        	Error Trace:	stopper_test.go:747
        	Error:      	Received unexpected error:
        	            	unexpected diff:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing.CheckRecordedSpans
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/test_utils.go:158
        	            	  | pkg/util/stop/stop_test_test.TestStopperRunAsyncTaskTracing
        	            	  | 	pkg/util/stop/stop_test_test/pkg/util/stop/stopper_test.go:747
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1259
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1581
        	            	Wraps: (2) unexpected diff:
        	            	  | --- exp
        	            	  | +++ got
        	            	  | @@ -1,5 +1,5 @@
        	            	  |  span: parent
        	            	  |      tags: _verbose=1
        	            	  |      span: async child same trace
        	            	  | -        tags: _verbose=1
        	            	  | +        tags: _unfinished=1 _verbose=1
        	            	  |          event: async 2
        	            	  |
        	            	  | recording:
        	            	  |      0.000ms      0.000ms    === operation:parent _verbose:1
        	            	  |      0.000ms      0.000ms    [async child same trace: {count: 1, duration 464µs, unfinished}]
        	            	  |      0.040ms      0.040ms        === operation:async child same trace _unfinished:1 _verbose:1
        	            	  |      0.160ms      0.119ms        event:pkg/util/stop/stop_test_test/stopper_test.go:718 async 2
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestStopperRunAsyncTaskTracing
--- FAIL: TestStopperRunAsyncTaskTracing (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStopperRunAsyncTaskTracing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17332

pkg/util/stop/stop_test.TestStopperRunAsyncTaskTracing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5880066?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5880066?buildTab=artifacts#/) on master @ [811ca1f37d3a286c66543888264da5105f9914b7](https://github.com/cockroachdb/cockroach/commits/811ca1f37d3a286c66543888264da5105f9914b7):


```
=== RUN   TestStopperRunAsyncTaskTracing
    stopper_test.go:747: 
        	Error Trace:	stopper_test.go:747
        	Error:      	Received unexpected error:
        	            	unexpected diff:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing.CheckRecordedSpans
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/test_utils.go:158
        	            	  | pkg/util/stop/stop_test_test.TestStopperRunAsyncTaskTracing
        	            	  | 	pkg/util/stop/stop_test_test/pkg/util/stop/stopper_test.go:747
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1439
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1571
        	            	Wraps: (2) unexpected diff:
        	            	  | --- exp
        	            	  | +++ got
        	            	  | @@ -1,5 +1,5 @@
        	            	  |  span: parent
        	            	  |      tags: _verbose=1
        	            	  |      span: async child same trace
        	            	  | -        tags: _verbose=1
        	            	  | +        tags: _unfinished=1 _verbose=1
        	            	  |          event: async 2
        	            	  |
        	            	  | recording:
        	            	  |      0.000ms      0.000ms    === operation:parent _verbose:1
        	            	  |      0.000ms      0.000ms    [async child same trace: {count: 1, duration 547µs, unfinished}]
        	            	  |      0.080ms      0.080ms        === operation:async child same trace _unfinished:1 _verbose:1
        	            	  |      0.472ms      0.392ms        event:pkg/util/stop/stop_test_test/stopper_test.go:718 async 2
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestStopperRunAsyncTaskTracing
--- FAIL: TestStopperRunAsyncTaskTracing (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStopperRunAsyncTaskTracing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/util/stop/stop_test.TestStopperRunAsyncTaskTracing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5907406?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5907406?buildTab=artifacts#/) on master @ [935198640534e087a4a86f3f05a7cae89c364acf](https://github.com/cockroachdb/cockroach/commits/935198640534e087a4a86f3f05a7cae89c364acf):


```
=== RUN   TestStopperRunAsyncTaskTracing
    stopper_test.go:747: 
        	Error Trace:	stopper_test.go:747
        	Error:      	Received unexpected error:
        	            	unexpected diff:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing.CheckRecordedSpans
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/test_utils.go:158
        	            	  | pkg/util/stop/stop_test_test.TestStopperRunAsyncTaskTracing
        	            	  | 	pkg/util/stop/stop_test_test/pkg/util/stop/stopper_test.go:747
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1439
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1571
        	            	Wraps: (2) unexpected diff:
        	            	  | --- exp
        	            	  | +++ got
        	            	  | @@ -1,5 +1,5 @@
        	            	  |  span: parent
        	            	  |      tags: _verbose=1
        	            	  |      span: async child same trace
        	            	  | -        tags: _verbose=1
        	            	  | +        tags: _unfinished=1 _verbose=1
        	            	  |          event: async 2
        	            	  |
        	            	  | recording:
        	            	  |      0.000ms      0.000ms    === operation:parent _verbose:1
        	            	  |      0.000ms      0.000ms    [async child same trace: {count: 1, duration 459µs, unfinished}]
        	            	  |      0.110ms      0.110ms        === operation:async child same trace _unfinished:1 _verbose:1
        	            	  |      0.520ms      0.410ms        event:pkg/util/stop/stop_test_test/stopper_test.go:718 async 2
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestStopperRunAsyncTaskTracing
--- FAIL: TestStopperRunAsyncTaskTracing (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStopperRunAsyncTaskTracing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/util/stop/stop_test.TestStopperRunAsyncTaskTracing [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6009290?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/6009290?buildTab=artifacts#/) on master @ [ce8bfd4811e789bdd4cf31640722517a5cf46237](https://github.com/cockroachdb/cockroach/commits/ce8bfd4811e789bdd4cf31640722517a5cf46237):


```
=== RUN   TestStopperRunAsyncTaskTracing
    stopper_test.go:747: 
        	Error Trace:	stopper_test.go:747
        	Error:      	Received unexpected error:
        	            	unexpected diff:
        	            	(1) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/util/tracing.CheckRecordedSpans
        	            	  | 	github.com/cockroachdb/cockroach/pkg/util/tracing/test_utils.go:158
        	            	  | pkg/util/stop/stop_test_test.TestStopperRunAsyncTaskTracing
        	            	  | 	pkg/util/stop/stop_test_test/pkg/util/stop/stopper_test.go:747
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1439
        	            	  | runtime.goexit
        	            	  | 	GOROOT/src/runtime/asm_amd64.s:1571
        	            	Wraps: (2) unexpected diff:
        	            	  | --- exp
        	            	  | +++ got
        	            	  | @@ -1,5 +1,5 @@
        	            	  |  span: parent
        	            	  |      tags: _verbose=1
        	            	  |      span: async child same trace
        	            	  | -        tags: _verbose=1
        	            	  | +        tags: _unfinished=1 _verbose=1
        	            	  |          event: async 2
        	            	  |
        	            	  | recording:
        	            	  |      0.000ms      0.000ms    === operation:parent _verbose:1
        	            	  |      0.000ms      0.000ms    [async child same trace: {count: 1, duration 513µs, unfinished}]
        	            	  |      0.082ms      0.082ms        === operation:async child same trace _unfinished:1 _verbose:1
        	            	  |      0.569ms      0.487ms        event:pkg/util/stop/stop_test_test/stopper_test.go:718 async 2
        	            	Error types: (1) *withstack.withStack (2) *errutil.leafError
        	Test:       	TestStopperRunAsyncTaskTracing
--- FAIL: TestStopperRunAsyncTaskTracing (0.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStopperRunAsyncTaskTracing.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/util/stop/stopper.go
+++ b/pkg/util/stop/stopper.go
@@ -478,8 +478,8 @@ func (s *Stopper) RunAsyncTaskEx(ctx context.Context, opt TaskOpts, f func(conte
 	// Call f on another goroutine.
 	taskStarted = true // Another goroutine now takes ownership of the alloc, if any.
 	go func() {
-		defer sp.Finish()
 		defer s.runPostlude()
+		defer sp.Finish()
 		defer s.recover(ctx)
 		if alloc != nil {
 			defer alloc.Release()

